### PR TITLE
C# Cleanup and refactoring

### PR DESCRIPTION
**Pull request by tamasvajk**
This PR is the first in multiple steps to remove nullability warnings. Currently when the context is created for CIL extraction, `null` is [passed](https://github.com/github/codeql/blob/b74911204a0c2f23c8c39be3b337d91566fa20fd/csharp/extractor/Semmle.Extraction.CIL/Entities/Assembly.cs#L139) as scope and compilation, which results in two nullability warnings. To overcome this, I think the best approach would be to have dedicated source and CIL extraction context classes. This PR only moves around functionalities. Most importantly, the `static` `Location` factories are moved to instance methods on the `Context`, which will later allow us to override them. 

[Differences job](https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/853/)

### CODE REVIEW

File: csharp/extractor/Semmle.Extraction/Context.cs [hunk 1]
```diff
@@ -225,14 +225,6 @@ public void PopulateAll()
...
 
-        /// <summary>
-        /// Create a new context, one per source file/assembly.
-        /// </summary>
-        /// <param name="e">The extractor.</param>
-        /// <param name="c">The Roslyn compilation.</param>
-        /// <param name="extractedEntity">Name of the source/dll file.</param>
-        /// <param name="scope">Defines which symbols are included in the trap file (e.g. AssemblyScope or SourceScope)</param>
-        /// <param name="addAssemblyTrapPrefix">Whether to add assembly prefixes to TRAP labels.</param>
         public Context(IExtractor e, Compilation c, TrapWriter trapWriter, IExtractionScope scope, bool addAssemblyTrapPrefix)
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction/Context.cs]: Why did you remove the doc comments?

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Method.AstLineCounter.cs [hunk 2]
```diff
@@ -4,9 +4,9 @@
 using Semmle.Util;
 using System.IO;
 
-namespace Semmle.Extraction.CSharp.Populators
+namespace Semmle.Extraction.CSharp.Entities
 {
-    public static class MethodExtensions
+    public abstract partial class Method
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction.CSharp/Entities/Method.AstLineCounter.cs:9]: I would prefer extension methods over `partial` classes.

**Review by hvitved — COMMENTED** [review]
Two questions, otherwise LGTM.

File: csharp/extractor/Semmle.Extraction/Context.cs (hunk 1 above)
Comment (tamasvajk) [reply csharp/extractor/Semmle.Extraction/Context.cs]: Because it was not matching the code. Also, one of the main goals is to refactor/split the `Context` class, so I expect more changes, which would make the comment even more outdated.

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Method.AstLineCounter.cs (hunk 2 above)
Comment (tamasvajk) [reply csharp/extractor/Semmle.Extraction.CSharp/Entities/Method.AstLineCounter.cs:9]: I could make `AstLineCounter` top level and public, and then there's no partial, and no extension method either. What do you think? I'm not convinced about all the extension methods that we have. They make sense when they extend Roslyn classes, but why have extension methods that extend classes that we directly control?

Also, in this particular case, `NumberOfLines` was an extension on `Context`, and it didn't use `Context` at all.

Comment (hvitved) [reply csharp/extractor/Semmle.Extraction.CSharp/Entities/Method.AstLineCounter.cs:9]: Then I prefer making `AstLineCounter` public.

File: csharp/extractor/Semmle.Extraction/Context.cs (hunk 1 above)
Comment (hvitved) [reply csharp/extractor/Semmle.Extraction/Context.cs]: OK

File: csharp/extractor/Semmle.Extraction.CSharp/Populators/AstLineCounter.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Semmle.Util;
+
+namespace Semmle.Extraction.CSharp.Populators
+{
+    internal class AstLineCounter : CSharpSyntaxVisitor<LineCounts>
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction.CSharp/Populators/AstLineCounter.cs:8]: Now `Method` no longer needs to be `partial`.

Comment (tamasvajk) [reply csharp/extractor/Semmle.Extraction.CSharp/Populators/AstLineCounter.cs:8]: Ahh, good point. Sorry...

**Review by hvitved — DISMISSED** [review]

**Comment by tamasvajk** [comment]
@hvitved I rebased this, and fixed the build issue after the rebase. Let's see if the tests pass, and if so, then it's ready to be merged.

**Review by hvitved — APPROVED** [review]
